src/BayesFilters/src/KFCorrection.cpp

```diff
@@ -0,0 +1,104 @@
+#include <BayesFilters/KFCorrection.h>
+
+using namespace bfl;
+using namespace Eigen;
+
+
+KFCorrection::KFCorrection(std::unique_ptr<LinearMeasurementModel> measurement_model) noexcept :
+    measurement_model_(std::move(measurement_model))
+{ }
+
+
+KFCorrection::KFCorrection(KFCorrection&& kf_correction) noexcept :
+    measurement_model_(std::move(kf_correction.measurement_model_))
+{ }
+
+
+KFCorrection::~KFCorrection() noexcept
+{ }
+
+
+MeasurementModel& KFCorrection::getMeasurementModel()
+{
+    return *measurement_model_;
+}
+
+
+void KFCorrection::correctStep(const GaussianMixture& pred_state, GaussianMixture& corr_state)
+{
+    /* Get the current measurement if available. */
+    bool valid_measurement;
+    Data measurement;
+    std::tie(valid_measurement, measurement) = measurement_model_->measure();
+
+    if (!valid_measurement)
+    {
+        corr_state = pred_state;
+        return;
+    }
+
+    /* Evaluate the predicted measurement if possible. */
+    bool valid_predicted_measurement;
+    Data predicted_measurement;
+    std::tie(valid_predicted_measurement, predicted_measurement) = measurement_model_->predictedMeasure(pred_state.mean());
+
+    if (!valid_predicted_measurement)
+    {
+        corr_state = pred_state;
+        return;
+    }
+
+    /* Evaluate the innovation if possible. */
+    bool valid_innovation;
+    Data innovation;
+    std::tie(valid_innovation, innovation) = measurement_model_->innovation(predicted_measurement, measurement);
+
+    if (!valid_innovation)
+    {
+        corr_state = pred_state;
+        return;
+    }
+
+    bool valid_covariance_matrix;
+    MatrixXd R;
+    std::tie(valid_covariance_matrix, R) = measurement_model_->getNoiseCovarianceMatrix();
+
+    if (!valid_covariance_matrix)
+    {
+        corr_state = pred_state;
+        return;
+    }
+
+    MatrixXd H = measurement_model_->getMeasurementMatrix();
+
+    /* Cast innovations once for all. */
+    MatrixXd innovations = any::any_cast<MatrixXd&&>(std::move(innovation));
+
+    /* Process all the components in the mixture. */
+    for (size_t i=0; i < pred_state.components; i++)
+    {
+        /* Evaluate the measurement covariance matrix
+           Py = H * Px * H' + R */
+
+        MatrixXd Py = H * pred_state.covariance(i) * H.transpose() + R;
+
+        /* Evaluate the Kalman Gain
+           K = Px * H' * (Py)^{-1} */
+
+        MatrixXd K = pred_state.covariance(i) * H.transpose() * Py.inverse();
+
+        /* Evaluate the filtered mean
+           x_{k}+ = x{k}- + K * (y - y_predicted) */
+
+        corr_state.mean(i) = pred_state.mean(i) + K * innovations.col(i);
+
+        /* Evaluate the filtered covariance
+           P_{k}+ = P_{k}- - K * Py * K' */
```

Comment:
Please standardize the way comments are placed before the code, either separeted by one or no empty line.